File: funcscript/block/KvcExpression.cs
using System.ComponentModel.Design;
using System.Data;
using funcscript.core;
using funcscript.model;
using System.Security.Cryptography;
using System.Text;
using funcscript.error;

namespace funcscript.block
{
    public class KvcExpression : ExpressionBlock
    {
        private class KvcExpressionProvider : IFsDataProvider
        {
            private readonly KvcExpression _parent;
            private readonly Dictionary<string, object> _valCache = new Dictionary<string, object>();
            public IFsDataProvider ParentProvider { get; }

            public bool IsDefined(string key)
            {
                return _parent.index.ContainsKey(key);
            }

            public KvcExpressionProvider(IFsDataProvider provider, KvcExpression parent)
            {
                this.ParentProvider = provider;
                _parent = parent;
            }
            String _evaluating = null;
            public object Get(string name)
            {
                if (_valCache.TryGetValue(name, out var val))
                    return val;
                if (_evaluating == null || name != _evaluating)
                {
                    if (_parent.index.TryGetValue(name, out var exp) && exp.ValueExpression != null)
                    {
                        _evaluating = name;
                        var v = exp.ValueExpression.Evaluate(this).Item1;
                        _evaluating = null;
                        _valCache[name] = v;
                        return v;
                    }
                }
                return ParentProvider.Get(name);
            }
        }

        public class KeyValueExpression
        {
            public String Key;
            public String KeyLower;
            public ExpressionBlock ValueExpression;
        }

        public class ConnectionExpression
        {
            public ExpressionBlock Source;
            public ExpressionBlock Sink;
            public ExpressionBlock Catch;
        }


        public IList<KeyValueExpression> _keyValues;
        public ExpressionBlock singleReturn = null;

        private Dictionary<string, KeyValueExpression> index;

        class ConnectionInfo
        {
            public object sink;
            public object vref;

            public CodeLocation location;

            protected void ThrowInvalidConnectionError(string msg)
            {
                throw new error.EvaluationException(location, new TypeMismatchError(msg));
            }
        }

        

        public String SetKeyValues(IList<KeyValueExpression> kv, ExpressionBlock retExpression,
            IList<ConnectionExpression> datConns,
            IList<ConnectionExpression> sigConns)
        {
            _keyValues = kv;
            this.singleReturn = retExpression;

            if (_keyValues == null)
                index = null;
            else
            {
                index = new Dictionary<string, KeyValueExpression>();
                foreach (var k in _keyValues)
                {
                    k.KeyLower = k.Key.ToLower();
                    if (this.index.ContainsKey(k.KeyLower))
                        return $"Key {k.KeyLower} is duplicated";
                    this.index.Add(k.KeyLower, k);
                }
            }

            return null;
        }

        public IList<KeyValueExpression> KeyValues => _keyValues;

        public override (object, CodeLocation) Evaluate(IFsDataProvider provider)
        {
            var evalProvider = new KvcExpressionProvider(provider, this);

            var kvc = new SimpleKeyValueCollection(null, this._keyValues
                .Select(kv => KeyValuePair.Create<string, object>(kv.Key,
                    evalProvider.Get(kv.KeyLower))).ToArray());
            

            if (singleReturn != null)
            {
                return singleReturn.Evaluate(evalProvider);
            }
            return (kvc, this.CodeLocation);
        }

        public override IList<ExpressionBlock> GetChilds()
        {
            var ret = new List<ExpressionBlock>();
            ret.AddRange(this.KeyValues.Select(x => x.ValueExpression));
            return ret;
        }

        public override string ToString()
        {
            return "Key-values";
        }

        public override string AsExpString(IFsDataProvider provider)
        {
            var sb = new StringBuilder();
            sb.Append("{\n");
            foreach (var kv in this.KeyValues)
            {
                sb.Append($"\t\n{kv.Key}: {kv.ValueExpression.AsExpString(provider)},");
            }

            if (this.singleReturn != null)
            {
                sb.Append($"return {this.singleReturn.AsExpString(provider)}");
            }

            sb.Append("}");
            return sb.ToString();
        }
    }
}

File: js_port/fs_js/src/block/KvcExpression.js
const { ExpressionBlock } = require('./ExpressionBlock');
const { SimpleKeyValueCollection } = require('../model/KeyValueCollection');
const { ensureTyped, makeValue } = require('../core/value');
const { FSDataType } = require('../core/fstypes');
const { FsDataProvider } = require('../core/dataProvider');

class KvcExpressionProvider extends FsDataProvider {
  constructor(parent, parentExpression) {
    super(parent);
    this.parentExpression = parentExpression;
    this.cache = new Map();
    this.evaluating = new Set();
  }

  get(name) {
    const lower = name.toLowerCase();
    const entry = this.parentExpression._index.get(lower);
    if (!entry) {
      return super.get(name);
    }
    if (this.cache.has(lower)) {
      return this.cache.get(lower);
    }
    if (this.evaluating.has(lower)) {
      return super.get(name);
    }

    this.evaluating.add(lower);
    try {
      const value = ensureTyped(entry.ValueExpression.evaluate(this));
      this.cache.set(lower, value);
      return value;
    } finally {
      this.evaluating.delete(lower);
    }
  }

  isDefined(name) {
    const lower = name.toLowerCase();
    if (this.parentExpression._index.has(lower)) {
      return true;
    }
    return super.isDefined(name);
  }
}

class KvcExpression extends ExpressionBlock {
  constructor() {
    super();
    this._keyValues = [];
    this.singleReturn = null;
    this._index = new Map();
  }

  SetKeyValues(kvExpressions, returnExpression) {
    this._keyValues = kvExpressions || [];
    this.singleReturn = returnExpression || null;
    this._index = new Map();
    for (const kv of this._keyValues) {
      const lower = kv.KeyLower || kv.Key.toLowerCase();
      if (this._index.has(lower)) {
        return `Key ${lower} is duplicated`;
      }
      kv.KeyLower = lower;
      this._index.set(lower, kv);
    }
    return null;
  }

  get KeyValues() {
    return this._keyValues;
  }

  evaluate(provider) {
    const scope = new KvcExpressionProvider(provider, this);
    const pairs = this._keyValues.map((kv) => [kv.Key, scope.get(kv.KeyLower)]);
    const collection = new SimpleKeyValueCollection(null, pairs);
    if (this.singleReturn) {
      return ensureTyped(this.singleReturn.evaluate(scope));
    }
    return makeValue(FSDataType.KeyValueCollection, collection);
  }

  getChilds() {
    const children = this._keyValues.map((kv) => kv.ValueExpression);
    if (this.singleReturn) {
      children.push(this.singleReturn);
    }
    return children;
  }

  asExpressionString(provider) {
    const parts = this._keyValues.map(
      (kv) => `${kv.Key}: ${kv.ValueExpression.asExpressionString(provider)}`
    );
    if (this.singleReturn) {
      parts.push(`return ${this.singleReturn.asExpressionString(provider)}`);
    }
    return `{ ${parts.join('; ')} }`;
  }
}

class KeyValueExpression {
  constructor() {
    this.Key = null;
    this.KeyLower = null;
    this.ValueExpression = null;
  }
}

module.exports = {
  KvcExpression,
  KeyValueExpression
};
